[노을] 미션3 : 체스판 초기화

sanhee opened the pull request:
안녕하십니까, 백엔드반 `노을` 입니다!! 👦
오늘도 리뷰 해주셔서 감사합니다 :)

미션3을 진행 하면서, 궁금한 점과 고민한 점을 정리하겠습니다.

<br>

#### 1. 커밋을 최대한 자세하게 작성해보았는데, 잘 작성했는지 모르겠습니다. 

#### 2. 흰색/흑색 Pawn의 생성, 개수확인 등 작은 단위의 테스트들을 만들어보고 싶었으나 `@Test` 마다 환경이 초기화 되는 것 같아, `@Test` 메소드 한 개에 검사하고자 하는 메소드를 한꺼번에 넣었습니다. 

   >- 미션2 리뷰: https://github.com/codesquad-members-2021/java-chess/pull/60#discussion_r575582823 
   >- 미션3 커밋링크:  26a0906fcdc74e5adb3a0ba181a330f70c792f22

#### 3. Pawn 열의 결과를 생성하는 `getWhitePawnsResult()`, `getBlackPawnsResult()` 메소드의 리팩토링을 어떻게 해야 할지 감이 오질 않습니다.  

   >- 좋은 방법이 떠오르지 않아, 단순히 메소드를 분리 해두긴 했는데 조언해주시면 감사하겠습니다.
   >- 커밋링크: d72224de95d629152bb30017e0b372592c6ba18a

#### 4. 체스를 출력하기 위해 2차원 리스트(`List<String[]> boardList`) 를 사용하였는데, 올바른 방식인지 궁금합니다.

 >- 커밋링크1: c142e1176304cec90d774790b341a891e4621f6f
 >- 커밋링크2: c191f00120b6d6617406e148812a3e79cab9c054

#### 5. initialize() 메소드에서 중복되는 리스트 삽입과정을, 반복문으로 리팩토링 하다보니 가독성이 더 떨어진 것 같습니다.

 >- 커밋링크: a898985b37104a9f0f07e69d00458951efd6b98b

#### 6. 색상 관리를 위해 enum 을 간단히 사용 중인데,  어느 부분에서 활용을 하면 좋을지 잘 모르겠습니다. ( enum value나 method를 활용해보고 싶습니다.)
```java
public enum UnitColor {
    BLACK, WHITE;
}
```

sanhee committed:
[Test] 호눅스님의 수업을 참고하여, 접근제한자를 제거하였습니다.

테스트 코드와 메소드에는 접근 제한자를 쓰지 않는게 일반적인 것을 몰랐습니다.

sanhee committed:
[Test] 리뷰를 반영하여, Board 객체를 BeforeEach를 통해 초기화

기존에 각각의 폰 마다 모듈단위의 테스트를 하고자, Board 객체를 static으로 선언하여
@Test 메소드간에 정보를 공유하였지만, 각 @Test가 순차적으로 실행되지 않고
독립된 환경에서 이루어짐을 알게되었습니다.
다만 현재 제 수준으로는 여러 개의 테스트를 모듈로 나눌 수 없어
한 테스트에 분리 된 검사 메소드를 담았습니다.

sanhee committed:
[feat] Pawn을 구분하기 위해, 색상에 따라 문자 지정

요구사항에 있는 검은색 Pawn은 대문자 P, 흰색 Pawn은 소문자 p를 반영하였습니다.
기본 생성자는 this() 생성자 함수 호출로, 인자가 없을 경우를 생각했습니다.
생성자 함수의 인자로 들어오는 enum Color 상수에 따라 모형을 Switch문의 조건에 따라 기물 모양을 지정해줍니다.

sanhee committed:
[Feat] 고정된 값인 Pawn의 최대 개수를 static 상수로 지정

Board 클래스의 initialize() 메소드에서 검은색/흰색 폰을 각각 8개씩 지정하는 요구사항을 수행중에
매직 넘버인 8을 상수로 지정해서 가독성을 높이고자 하였습니다.

sanhee committed:
[Test] Pawn 색에 따른 출력 문자 유효성 검사

(*참고*) 요구사항에 있는 것처럼 Pawn을 생성할 때, 두 번째 인자로 출력할 문자를 지정하지 않았습니다.
- Pawn의 생성자 함수에서 색상에 따라, 문자가 자동 지정됩니다.

sanhee committed:
[Feat] Pawn의 기호를 상수로 지정 및 기호 리턴 메소드 구현

자주 사용되는 색상별 폰의 모양(p/P)를 static 상수로 지정하였습니다.
Pawn 객체의 기호를 리턴하는, getRepresentation() 메소드를 구현하였습니다.

sanhee committed:
[Test] 생성된 흰색/검은색 Pawn 열의 결과 검증

initialize() 메소드에서 흰색 Pawn 열의 결과가 pppppppp
검은색 Pawn 열의 결과가 PPPPPPPP 인지를 검증하는 테스트 입니다.

sanhee committed:
[Test] initialize() 메소드 기대값 문자열로 수정

실수로 문자열로 지정하지 않아 수정하였습니다.

sanhee committed:
[Feat] 생성된 Pawn 객체의 기호 문자열을 리턴하는 메소드 정의

# pawns 리스트에 Pawn 객체가 차례대로 삽입되어, 문자열을 만드는 반복문의 과정에서 매직넘버가 포함되었습니다.

* 메소드: getWhitePawnsResult()
 - 흰색 Pawn의 기호 문자(p)를 StringBuilder로 모아 리턴합니다.

* 메소드: getBlackPawnsResult()
 - 흑색 Pawn의 기호 문자(P)를 StringBuilder로 모아 리턴합니다.

sanhee committed:
[Feat] 체스판에서 사용 할 유닛 상수 정의

* String defaultUnitRep
 : Pawn이 아닌 기본 오브젝트 (.)

* Pawn[] whitePawn
 : 흰색 폰 객체 배열로 최대 개수인 8개로 선언되었습니다.

* Pawn[] blackPawn
 : 흑색 폰 객체 배열로 최대 개수인 8개로 선언되었습니다.

sanhee committed:
[Chore] defaultUnitRep 상수의 불필요한 static 제거

sanhee committed:
[Feat] 유닛이 위치 할 체스판을 2차원 리스트로 정의

* List<String[]> boardList

2차원 리스트로 정의하여, 각 포지션마다 오브젝트를 삽입할 수 있게 하였습니다.

sanhee committed:
[Feat] 체스판의 유닛들을 초기화 하는 initialize() 메소드 정의

1. 이전에 정의해둔 각 Color의 Pawn 객체배열을 초기화 해줍니다.

2. 2차원 리스트, boardList에 체스 유닛들을 차곡차곡 삽입해줍니다.

sanhee committed:
[Chore] 의미 없는 변수명 정정

Pawn 객체의 기호 정보를 모으는 StringBuilder의 이름을 변경했습니다.

sanhee committed:
[Refactor] 중복되는 Pawn 객체 배열 초기화 arrayPawnInit() 메소드로 분리

Board 클래스의 initialize() 메소드

sanhee committed:
[Refactor] 중복되는 boardList 삽입과정 반복문 사용

중복되는 라인들을 반복문으로 줄였습니다.

** 다만, 가독성이 반복문을 사용하기전이 더 좋은 것 같습니다.

sanhee committed:
[Refactor] 중복되는 Pawn 열의 결과를 생성하는 코드 분리

하나의 반복문으로 압축할 아이디어가 떠오르지 않아,

switch문을 활용해 메소드 분리를 하였습니다.

sanhee committed:
[Test] 체스 판의 결과를 콘솔에 출력

sanhee committed:
[Feat] 체스 판의 결과를 출력하는 print() 정의

StringBuilder를 활용해, 체스판의 List 정보를 저장후 String으로 변환하여 리턴합니다.

sanhee committed:
[Feat] 체스게임 시작 및 종료 구현

ChessGame 클래스를 생성하였습니다.
Scanner를 통해 명령어를 입력 받아, 체스 게임을 진행할 수 있습니다.

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +25,67 @@ public int size() {
+    }
+
+    public String appendPawnRep(UnitColor color) {
+        StringBuilder PawnLine = new StringBuilder();
```
변수 이름은 camelCase로 써주세요~

sanhee replied:
와!!!! 빠르고 세심한 리뷰 너무 감동 받았습니다 👍
리뷰 반영해서 수정했습니다!

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +25,67 @@ public int size() {
+        }
+
+        return PawnLine.toString();
+    }
```
요 로직 뿐 아니라 전반적으로 for i 문이 많은 것은 객체를 별로 활용하지 못하고 있다는 사인이 될 수 있습니다. 왜냐하면 포문 안에 `int i`가 등장한 이유는 `i`를 가지고 리스트의 인덱스로 접근하겠다는 의도가 있기 때문인데요~ 보통 이렇게 되면 데이터를 꺼내오는 식의 사고를 주로 하게 되어 객체 중심적으로 짜기 어렵게 될 수도 있습니다. (제 생각입니다.ㅋㅋ)

for문을 사용할때 for each 문 또는 스트림을 활용해보세요

저라면 가지고 있는 pawns를 이용하여 충분히 간결하게 짤 수 있을 것 같네요
```java
public String appendPawnRep(UnitColor color) {
  StringBuilder pawnLine = new StringBuilder();
  for (pawn : pawns) {
    // 색상이 같으면 append
    if (pawn.isColor(color)) pawnLine.append(pawn.getRepresentation());
  }
  return pawnLine.toString();
}
```

sanhee replied:
와...✨ 저는 이런 생각을 도저히 하지 못했었는데..
새로운 사실과 제 무지함(?)에..정말 충격 받았습니다. 
팁 감사합니다.  stream으로도 한번 도전 해보겠습니다.

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +25,67 @@ public int size() {
+            arrayPawns[i] = new Pawn(color);
+            pawns.add(arrayPawns[i]);
+        }
+    }
```
클래스 내부에서만 쓰이는 메서드이므로 private으로 선언해주세요

sanhee replied:
감사합니다, 수정했습니다!

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -1,13 +1,18 @@
+    private final String defaultUnitRep = "........";
+    private final Pawn[] whitePawn = new Pawn[Pawn.MAX_SPAWN_NUMBER];
+    private final Pawn[] blackPawn = new Pawn[Pawn.MAX_SPAWN_NUMBER];
+    private final List<String[]> boardList = new ArrayList<>();
```
멤버 변수를 3개 이하로 줄여보세요. (충분히 가능합니다!)

보드를 출력하는데 필요한 최소한의 정보는
```java
private final List<Pawn> pawns = new ArrayList<Pawn>();
private final String defaultUnitRep = "........";
```

이정도만 있어도 충분할 것 같습니다. 위치에 대한 정보가 필요하다면 Pawn 안으로 넣을 수도 있고,
위치,Pawn 을 감싼 새로운 클래스가 필요할 수도 있을 것 같구요~

특히나 whitePawn, blackPawn 은 arrayPawnInit 의 로컬변수로 선언이 되어도 될것으로 보이네요!

sanhee replied:
리뷰 감사합니다 ㅎㅎ
우선 whitePawn, blackPawn을 initialize() 메소드의 지역변수로 만들고 push 하였습니다.
boardList는 좀 더 생각 한 후 개선하겠습니다!

sanhee replied:
boardList를 print() 내부로 이동시켰습니다!

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +25,67 @@ public int size() {
+        arrayPawnInit(whitePawn, UnitColor.WHITE);
+        arrayPawnInit(blackPawn, UnitColor.BLACK);
+
+        for (int i = 0; i < 8; i++) {
+            if (i == 1 || i == 6) {
+                boardList.add(getBlackPawnsResult().split(""));
+            } else {
+                boardList.add(defaultUnitRep.split(""));
+            }
+        }
```
그런데..  boardList가 꼭 initialize할때 초기화 되어야 할까요? 
initialize할때는 기물(현재는 pawn)만 잘 셋팅을 하고, Print 할때 그때그때 로직을 돌려도 될것 같은데요~

지금처럼 폰 셋팅할때 한번에 프린트될 내용까지 초기화를 하면, 게임 도중 폰 색깔이 바뀔 때 정합성 문제가 생길 수 있고
프린트 로직은 다른 기능으로 분리하기도 어렵습니다.

요지는 boardList 또한 꼭 가지고 있어야 할 멤버변수인지 고민해보시고 ㅎㅎ
Board가 어떤 데이터가 아닌 꼭 필요한 객체는 무엇인지? (추상적이네요..) 그리고 그 객체로 뭘 할 수 있을지? 를 고민해주시면 좋을 것 같습니다

sanhee replied:
그러고보니 기존 상태는 프린트를 하면 초기 상태의 체스 보드판만을 출력해서 게임 진행이 불가능했겠네요.. 
initialize 에 넣었던 boardList가 게임 도중에 문제가 생길거란걸 미처 생각 못했습니다.😥 

추가적으로, boardList 또한 꼭 필요한 멤버가 아닌 것 같아서, print() 메소드의 로컬변수로 변경했습니다.

섬세한 리뷰 감사합니다!🤠

LarryJung commented on src/main/java/net/sanhee/pieces/Pawn.java:
```diff
@@ -2,17 +2,35 @@
 
 public class Pawn {
+    public static final int MAX_SPAWN_NUMBER = 8;
+    public static final char BLACK_REPRESENTATION = 'P';
+    public static final char WHITE_REPRESENTATION = 'p';
```
enum 을 좀 활용해보자면..
```java
private static final Map<UnitColor, char> colorMap;

...

public UnitColor getColor() {
  return this.colorMap.get(this.color);
}
```

이렇게 할 수도 있겠네요~

sanhee replied:
```java
    private static final Map<UnitColor, Character> colorMap = new HashMap<UnitColor, Character>() {{
        put(UnitColor.BLACK, 'P');
        put(UnitColor.WHITE, 'p');
    }};

   public char getShape() {
        return colorMap.get(this.color);
    }
```

와 같이 구현을 했는데, 지금 단계에서는 final로 선언해둔게 좀 더 직관적인 것 같아서 커밋을 하지 않았는데 제가 잘못 구현한걸까요??

LarryJung replied:
아~ 아니요 ㅎㅎ 정답은 없으니 나중에 더 수정이 필요하다 싶으면 계속 고치시면 되겠습니다!

LarryJung commented on src/main/java/net/sanhee/pieces/Pawn.java:
```diff
@@ -2,17 +2,35 @@
+    public static final char WHITE_REPRESENTATION = 'p';
+
     private final UnitColor color;
+    private final char shape;
```
Board와 유사하게 말씀드리면.. 
1. shape가 미리 선언이 되는 것과
2. getShape() 할때 그때 로직을 돌리기

두 가지 방식으로 짤 수 있을 것 같고 위에 map 예제 드린 것과 한번 비교해보세요!

sanhee replied:
map 예제를 하기전에 먼저 할 수 있는 코드를 짜보았습니다!

```java
public char getShape(){
        switch (this.color) {
            case BLACK:
                return BLACK_REPRESENTATION;
            case WHITE:
            default:
                return WHITE_REPRESENTATION;
        }
    }
```

LarryJung commented on src/test/java/BoardTest.java:
```diff
@@ -1,41 +1,62 @@
+class BoardTest {
 
-    static public Board board = new Board();
+    Board board;
+    int totalPawnCnt;
```
특별한 목적이 없다면 private으로 선언해주세요.

LarryJung replied:
그리고 initialize() 테스트에서 하셨던 것처럼 Board를 꼭 테스트 클래스의 멤버 변수로 필요는 없을 것 같습니다~ createPawn 테스트 안에서 직접 로컬변수로 선언해서 테스트환경을 더 좁게 가둬두는것도 좋습니다.ㅎㅎ

sanhee replied:
테스트코드에서 클래스와 메소드에 접근제어자를 쓰지 않는 것을 
변수에도 해당하는것으로 착각했었습니다...
수정하였습니다. 감사합니다😊

LarryJung commented on src/main/java/net/sanhee/chess/ChessGame.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.Scanner;
+
+public class ChessGame {
+    public static void main(String[] args) {
+
+        System.out.print("체스게임(start/end) > ");
+        Scanner scanner = new Scanner(System.in);
+
+        boolean endGame = false;
+        while (!endGame) {
+            String cmd = scanner.nextLine();
+            switch (cmd) {
+                case "start":
+                    Board board = new Board();
+                    board.initialize();
+                    System.out.println(board.print());
+                    break;
+                case "end":
+                    System.out.println("> 게임이 종료되었습니다.");
+                    endGame = true;
+                    break;
+                default:
+                    System.out.println("start/end 중 하나를 입력하세요.");
+                    break;
+            }
+
+        }
+        scanner.close();
+    }
+}
```
👍

sanhee replied:
감사합니다😁 
현재 바로 고칠 수 있는 간단한 수정사항부터 우선적으로 진행 하고 있습니다.👀

LarryJung reviewed (commented):
3단계부터 요구사항이 확 복잡해져서 고민을 많이하신 게 느껴집니다!
잘 하셨지만 더 간결하게 짤 수 있는 방법에 대해서 좀 더 학습해 보실 것을 권해드립니다.

sanhee committed:
[chore] PawnLine 변수명 camelCase로 변경

LarryJung님의 리뷰를 반영하여, 수정하였습니다👍

sanhee committed:
[chore] 클래스내에서만 사용하는 arrayPawnInit()의 접근제어자를 private로 변경

LarryJung의 리뷰를 반영하였습니다👍

sanhee committed:
[test] 접근제어자 추가와 불필요한 멤버변수 로컬변수로 변환

1. 테스트코드에서 클래스와 메소드에 접근제어자를 쓰지 않는 것을 변수에도 해당하는것으로 착각했었습니다.
2. LarryJung님의 리뷰를 반영하여, 테스트환경을 더 좁게 만들기 위해 기존멤버변수로 선언되어있던 Board를 로컬변수로 수정했습니다.

sanhee committed:
[refactor] Board클래스의 appendPawnRep()에서 불필요한 for문 제거

LarryJung님의 리뷰를 반영하였습니다.
이후에는 스트림으로 개선할려고 합니다.

sanhee committed:
[refactor] Board 클래스에서 불필요한 멤버변수 로컬변수로 변환

LarryJung님의 리뷰를 반영하여,
initialize()에서만 사용하는 whitePawn, blackPawn 을 로컬변수로 변환했습니다.

sanhee committed:
[refactor] 정합성 문제를 일으킬 boardList 갱신을 print()에서 담당하는것으로 변경

- boardList를 클래스 멤버변수에서 print()의 로컬변수로 변경했습니다.

- 기존 boardList는  initialize 할때만 갱신되어, 초기 상태의 체스 보드판만을 출력하는 문제가 있었습니다.

sanhee committed:
[refactor] 가독성을 위해 boardList를 추가하는 과정을 하드코딩으로 변경

- 중복되는 라인이 존재해 반복문과 조건문을 통해 중복을 제거했지만, 가독성이 더 떨어져 하드코딩으로 회귀하였습니다.

sanhee committed:
[refactor] Pawn 클래스의 모양 지정 추가 메소드로 분리

- getShape()메소드가 호출 될때 모양을 지정하고, 리턴하도록 변경했습니다.

sanhee commented:
리뷰해주신 사항 반영하면서 `Resolve Conversation` 버튼을 눌렀는데,
제가 임의로 눌러도 되는게 맞을까요??

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +22,61 @@ public int size() {
+        return appendPawnRep(UnitColor.BLACK);
+    }
+
+    public String appendPawnRep(UnitColor color) {
+        StringBuilder pawnLine = new StringBuilder();
+        for (Pawn pawn : pawns) {
+            if (pawn.isColor(color)) {
+                pawnLine.append(pawn.getRepresentation());
+            }
+        }
+        return pawnLine.toString();
+    }
```
이전보다 이렇게 하는게 더 보기가 좋은것 같네요 저는 .^^

sanhee replied:
저도 그렇게 생각합니다!! 훨씬 간결해지고 이해하기 쉬운 코드가 된 것 같습니다. ㅎㅎ

LarryJung commented on src/main/java/net/sanhee/chess/Board.java:
```diff
@@ -20,4 +22,61 @@ public int size() {
+        return pawnLine.toString();
+    }
+
+    public void initialize() {
+        Pawn[] whitePawn = new Pawn[Pawn.MAX_SPAWN_NUMBER];
+        Pawn[] blackPawn = new Pawn[Pawn.MAX_SPAWN_NUMBER];
+        arrayPawnInit(whitePawn, UnitColor.WHITE);
+        arrayPawnInit(blackPawn, UnitColor.BLACK);
+    }
+
+    private void arrayPawnInit(Pawn[] arrayPawns, UnitColor color) {
+        for (int i = 0; i < Pawn.MAX_SPAWN_NUMBER; i++) {
+            arrayPawns[i] = new Pawn(color);
+            pawns.add(arrayPawns[i]);
+        }
+    }
+
```
이 부분도 조금 리팩토링을 해보자면

```java
public void initialize() {
        arrayPawnInit(UnitColor.WHITE);
        arrayPawnInit(UnitColor.BLACK);
    }

private void arrayPawnInit(UnitColor color) {
    for (int i = 0; i < Pawn.MAX_SPAWN_NUMBER; i++) {
        pawns.add(new Pawn(color));
    }
}
```

이렇게 할 수 있지 않을까요? 변수는 적을수록 좋다는 말도 있으니(다 그런건 아닙니다만..) 기억해주세요!

sanhee replied:
이 부분도 전혀 생각치 못했는데, 코드를 짤 때 조금 더 신경써서 짜야겠습니다😂
꼭 필요없는 변수를 사용해서 낭비하고 있었네요.
리뷰해주신 내용 반영해서 push 했습니다!!

정말 섬세한 리뷰 너무 감사드립니다..감동입니다 😀

sanhee committed:
[refactor] Board 클래스의 initialize() 메소드의 불필요한 변수 제거

- LarryJung님의 리뷰를 반영하여, 기존 whitePawn, blackPawn 은 사실상 temp 변수여서 없어도 되는 변수로 제거하였습니다.

LarryJung reviewed (approved):
네 확인했습니다! 수고하셨습니다.^^ 다음단계 준비하시죠~

sanhee commented:
죄송합니다...실수로 close 눌렀네요..

다음부터는 주의하겠습니다.
